ecleankernel/layout/blspec.py

```diff
@@ -45,19 +46,66 @@ def __init__(self,
         for path in ("etc/kernel/entry-token", "etc/machine-id"):
             try:
                 with open(root / path) as f:
-                    kernel_id = f.read().strip()
+                    self.kernel_id = f.read().strip()
                 break
             except FileNotFoundError:
                 pass
         else:
             raise LayoutNotFound("/etc/machine-id not found")
 
-        for d in self.potential_dirs:
-            self.bootdir = root / d / kernel_id
-            if self.bootdir.is_dir():
-                return
-        else:
-            raise LayoutNotFound(f"/boot/[EFI/]{kernel_id} not found")
+        bootctl = subprocess.check_output(["bootctl", "-x"], text=True)
+        try:
+            bootdir = root / bootctl.removeprefix('/').removesuffix('\n')
+        except AttributeError:
+            raise LayoutNotFound(f"bootctl failed to find the XBOOTLDR or ESP partition: {bootctl}")
+
+        # Type 1 entries (linux+initrd) are in $BOOT/ENTRY-TOKEN/KERNEL-VERSION/
+        self.blsdir = bootdir / self.kernel_id
+
+        # Type 2 entries (uki's) are in $BOOT/EFI/Linux/ENTRY-TOKEN-KERNEL-VERSION.efi
+        self.ukidir = bootdir / "EFI" / "Linux"
+
+        if not self.blsdir.is_dir() and not self.ukidir.is_dir():
+            raise LayoutNotFound(f"Neither /{bootdir}/{self.kernel_id}, nor /{bootdir}/EFI/Linux found or empty")
```

Comment:
It doesn't seem that you're checking for the "or empty" as the message suggests.

Reply:
Oops that is a left over from some earlier version.